---
title: "report.rmd"
output: html_document
author: David Lenci
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source("main.R")
```

First, I import the necessary gene expression data, ensuring the format of the resulting
dataframe is optimal to begin filtering the data. In this instance I made the columns genes and the rows samples.

```{r}
expr_mat <- read_expression_table('/usr4/bf527/dlenci/data/bf528_data/example_intensity_data.csv')
meta_data <- read_meta_table('/usr4/bf527/dlenci/data/bf528_data/proj_metadata.csv')
knitr::kable(head(expr_mat[c(1:5)]))
```
Next, I perform the necessary filters. First, I determine the percentage of data points in a column the are greater than log2(15). If that percentage is less than 20 I drop that column from the data. Then, I perform the chisquare filter on the data using a p value of 0.01 and a degree of freedom of 34 (N-1, where N = number of samples). Lastly, I calculate the coefficient of variation for each row and drop any row that has a CF < 0.186 (where CF = SD/Mean).

This returns a dataframe with less than 1000 genes compared to our starting value of 54,000 genes.


```{r}
filtered_data <- first_filter(expr_mat) %>%
  sec_filter() %>%
  third_filter()
knitr::kable(head(filtered_data[c(1:5)]))
```

Next, I perform the clustering. Before this, I use the agnes module to determine the type of hierarchical clustering that will return the tightest clustering. The below function will generate agglomerative coefficients using our data and the varying types of clustering available. An AC closer to 1 represents a better coefficient.

From this we see the 'ward' clustering will give us the best results.

```{r}
best_method(filtered_data)
```

Now, using the 'ward' clustering type I perform the clustering and plot it using ggplot.

```{r}
clusters <- clustering(filtered_data)
```

Next, I generate a heat_map using the heatmap function and attach a color bar to visualize
the different clusters.

```{r}
heat_map(filtered_data)
```
Lastly, I performed the welch t test across the two clusters for each gene, and filtered out any 
differentially expressed genes with p-value < 0.05.

```{r}
diff_exp_tab <- welch_t(filtered_data, clusters)
knitr::kable(head(diff_exp_tab[c(1:11)]))

```